beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -1849,13 +1861,119 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         Ok(())
     }
 
+    fn prune_abandoned_forks(
+        &self,
+        old_finalized_block_hash: SignedBeaconBlockHash,
+        new_finalized_block_hash: SignedBeaconBlockHash,
+        new_finalized_slot: Slot,
+    ) -> Result<(), Error> {
+        let old_finalized_slot = self
+            .get_block(&old_finalized_block_hash.into())?
+            .ok_or_else(|| Error::MissingBeaconBlock(old_finalized_block_hash.into()))?
+            .message
+            .slot;
+
+        // Collect hashes from new_finalized_block up to old_finalized_block
+        let newly_finalized_blocks: HashSet<SignedBeaconBlockHash> = HashSet::from_iter(
+            ParentRootBlockIterator::new(&*self.store, new_finalized_block_hash.into())
+                .take_while(|(block_hash, _)| *block_hash != old_finalized_block_hash.into())
+                .map(|(block_hash, _)| block_hash.into()),
+        );
+
+        // We don't know which blocks are shared among abandoned chains, so we buffer and delete
+        // everything in one fell swoop.
+        let mut abandoned_blocks: HashSet<SignedBeaconBlockHash> = HashSet::new();
+        let mut abandoned_states: HashSet<(Slot, BeaconStateHash)> = HashSet::new();
+        let mut abandoned_heads: HashSet<Hash256> = HashSet::new();
+
+        for (head_hash, head_slot) in self.heads() {
+            let mut potentially_abandoned_head: Option<Hash256> = Some(head_hash);
+            let mut potentially_abandoned_blocks: Vec<(
+                Slot,
+                Option<SignedBeaconBlockHash>,
+                Option<BeaconStateHash>,
+            )> = Vec::new();
+
+            for result in
+                SlotBlockStateIterator::from_block(Arc::clone(&self.store), head_hash.into())?
+            {
+                let (slot, is_skipped_slot, block_hash, state_hash) = result?;
+
+                if slot <= old_finalized_slot {
+                    // We must assume here any candidate chains include old_finalized_block_hash,
+                    // i.e. there aren't any forks starting at a block that is a strict ancestor of
+                    // old_finalized_block_hash.
+                    break;
+                }
+                if is_skipped_slot {
+                    if newly_finalized_blocks.contains(&block_hash) {
```

Comment:
Is it correct to push twice to the same vec?

Reply:
You mean the repeated code for both cases?  Yeah, the if-else can be collapsed sure.

Reply:
yes, this is what i meant